zmq_addon.hpp

```diff
@@ -469,6 +469,79 @@ class multipart_t
         return true;
     }
 
+    // Return single part message_t encoded from this multipart_t.
+    // This encodes in a manner that should be compatible with the
+    // packing done by CZMQ's zmsg_encode().
+    message_t encode() const {
+        size_t mmsg_size = 0;
+        for (auto& part : *this) {
+            size_t part_size = part.size();
+            size_t count_size = 5;
+            if (part_size < 255) {
+                count_size = 1;
+            }
+            mmsg_size += part_size + count_size;
+        }
+
+        message_t encoded(mmsg_size);
```

Comment:
Hm, I don't think so. It's a corner case, but why should it be invalid to have a 0-part multi-part message?

But I don't know if `zmsg_encode` allows that. It should be done like that function does, since the goal is to be compatible.

@brettviren Please double-check and confirm this. In any case, please add a test case for this corner case.

Reply:
So far I have yet to require sending an empty multipart-encoded in any higher-level protocols I've used.  Certainly, if it can be done it will be done eventually.

I think the proper thing is for an empty `multipart_t` to encode to the default (empty) `message_t` and vice versa. 

I'll double check what CZMQ does here and add some tests for this case.

Reply:
Indeed CZMQ encodes an empty multipart into an empty single part.  Tests for this case are added.